### ISSUE TITLE
VariableDeclarationUsageDistance and timing code

### ISSUE THREAD

**Issue by mebigfatguy** [checkstyle/checkstyle#21366]
It would be nice if timing code was not reported for this detector

long start = System.currentTimeMillis();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
stuff();
long timing = System.currentTimeMillis() - start;

**Comment by romani** [comment]
Please report issue by issue template.

I think in this case you would need suppression usage by variable name.

**Comment by romani** [comment]
@lithops-zty, what you can recommend here ?

**Comment by lithops-zty** [comment]
It is definitely possible to add a new suppression rule in the check (possibly also a property that turns it on) just like we did for initialization sequences. The difficulty is how we can define a 'timing variable' that is most accurate. Two approaches I can think of:
1. Suppress when name of violating variable contains 'time' or is 'start'
2. Suppress when the violation variable is 
    1. initialized by `System.currentTimeMillis()`, or
    2. initialized by `System.nanoTime()`, or
    3. of type `Instant`.

**Comment by romani** [comment]
User should suppress by variable name. Can we do xpath suppression?

**Comment by lithops-zty** [comment]
Yes. 
```xml
  <suppress-xpath
          checks="VariableDeclarationUsageDistance"
          files=".*"
          query="//VARIABLE_DEF[IDENT[
              contains(@text, 'time')
              or @text = 'start'
          ]]"/>
```
should do the job.

Alternatively, we can set `start` to `final` and by default it will exempted from check.

**Comment by romani** [comment]
If you put this xpath to https://checkstyle.org/filters/suppressionxpathsinglefilter.html#SuppressionXpathSingleFilter
And locate such suppression filter close to your Check config. It will be good pair that does what you want.

**Comment by mebigfatguy** [comment]
i would really like the suppression to be the stuff on the right hand side of a variable assignment, not the variable name itself, is that possible?

hmm, i guess another option is to create a CloseableTimer class, instead

**Comment by romani** [comment]
By xpath you can check right side also, just need to go up from variable and move down on other branch of `=`.
If share all by issue template , you would get more precise answer.

**Comment by lithops-zty** [comment]
```xml
    <suppress-xpath files=".*"
              checks="VariableDeclarationUsageDistance"
              query="//VARIABLE_DEF[
                  ./ASSIGN/EXPR/METHOD_CALL/DOT/IDENT[@text='System']
                  and ./ASSIGN/EXPR/METHOD_CALL/DOT/IDENT[@text='currentTimeMillis' or @text='nanoTime']
                  or ./TYPE/IDENT[@text='Instant']
              ]"/>
```
This is XPath that checks RHS of assignment or type matches `Instant`.
You can remove the `Instant` check if it's not needed.

**Comment by abhinav-phi** [comment]
Hi! I'd like to work on this.
Plan: I'll investigate the VariableDeclarationUsageDistance behavior with timing code as reported, fix the distance calculation/edge case, and add input test cases.
Could you please assign this to me? Thanks!

/assign take

**Comment by romani** [comment]
@abhinav-phi , for new contributors we recommend
Please [read](https://github.com/checkstyle/checkstyle/blob/master/.github/CONTRIBUTING.md#getting-started) and watch playlist videos at [Starting_Development](https://github.com/checkstyle/checkstyle/blob/master/docs/BEGINNING_DEVELOPMENT.md#starting-development).

**Comment by romani** [comment]
@mebigfatguy , I think we provided you good workaround.
I am closing issue.
discussions can continue if required.
